### ISSUE TITLE
THF: Formula T0A_Defined_Constant true not supported

### ISSUE THREAD

**Issue by eugenk** [spechub/Hets#1544]
THF is not supported completely:

```
$ wget https://gist.githubusercontent.com/eugenk/a8a82fa2aa0f82bb661a/raw/7f48f4bc4ca332c919588ffa53d9685232f51c2d/AGT0271.het
$ hets AGT0271.het
Analyzing file AGT0271.het as library AGT0271
logic THF
Analyzing spec AGT0271?AGT0271
/private/tmp/deleteme/AGT0271.het:5.1-476.80: *** Error:
Formula T0A_Defined_Constant true not supported!
hets-bin: user error (Stopped due to errors)
```

There is even a `fixme` comment in the code (THF/Poly.hs:298).

This is quite important for my master's thesis. I can't think of a workaround for this.

**Comment by sternk** [comment]
I probably wasn't sure how to handle this. Can you explain to me how these should be typed?

**Comment by eugenk** [comment]
Actually, I'm not sure either. I just took those files from the [TPTP](http://www.cs.miami.edu/~tptp/) distribution ([TPTP-v6.2.0.tgz](http://www.cs.miami.edu/~tptp/TPTP/Distribution/TPTP-v6.2.0.tgz)) and want to evaluate them myself.

**Comment by sternk** [comment]
It seems to be about this definition
>thf(mtrue,definition,
>    ( mtrue
>    = ( ^ [W: $i] : $true ) )).
But I don't know the syntax well enough to immediately see how to understand it.

**Comment by tillmo** [comment]
@sternk, we can ask Chris Benzmüller. Could you please formulate a question, then I can mail him it.

**Comment by sternk** [comment]
I believe that this gives the kind of the type $true. I would like to know how to read the definition i.e. what the kind is in this case and (if possible) another example to make this type of definition clear to me.

**Comment by cbenzmueller** [comment]
"$true": is the logical constant for Truth in TPTP. What happens here is that a new, "lifted" constant symbol "mtrue" is defined, which for all arguments W of type $i returns the logical constant symbol $true.
This definition is part of a possible world lifting of logical connectives to operate on possible worlds (the objects of type $i).

Best, Christoph

**Comment by sternk** [comment]
Ok. Is there any list of the "built-in" constants of TPTP?

**Comment by cbenzmueller** [comment]
Everything should be in here: http://www.cs.miami.edu/~tptp/TPTP/SyntaxBNF.html

**Comment by sternk** [comment]
Yeah right - I must have overlooked that. Thanks for the help!

**Comment by tillmo** [comment]
@sternk, any progress here?

**Comment by sternk** [comment]
I plan on getting this done this week.

**Comment by tillmo** [comment]
good!

**Comment by eugenk** [comment]
I have created a (temporarily available) git repository with over 3000 of such files for the thesis:
```
git clone https://gitlab.informatik.uni-bremen.de/eugenk/THF-embedded-in-Het.git
```
This can serve as a testbed for your implementation.

I ran it with
```
find . -iname *.het | xargs -I % hets -o xml --full-signatures -a none -v2 -O /tmp/ +RTS -K1G -RTS --full-theories -A %
```

There is also another error type:
```
./ALG/ALG247-2.p.het:597.7-601.44:
*** Error:
Different type constructor!  - (In-)Equality requires (pushprop_p_and_p_prime) : (term > subst > (term > $o) > (term > $o) > $o) and ((^ [A : term, M : subst, P : term > $o,
    Q : term > $o] : ! [X : term] : ((Q @ X) <=> (P @ (sub @ X @ (push @ A @ M)))))) : (term > $o > term > $o > subst > term > $o to have the same type
hets-bin: user error (Stopped due to errors)
```

**Comment by sternk** [comment]
This week and weekend were far more busy than I had hoped for. I'll hopefully be able to finish work on this issues coming Tuesday.

**Comment by tillmo** [comment]
looking forward to this...

**Comment by sternk** [comment]
I have now fixed the described errors and am waiting if all the files go through

**Comment by tillmo** [comment]
good! Have the files gone through? @eugenk, perhaps you could also have a look.

**Comment by eugenk** [comment]
I'm running hets right now now the files on te.iws.cs.ovgu.de and it's looking good.

**Comment by sternk** [comment]
I encountered another issue and am working in resolving it

**Comment by eugenk** [comment]
Which issue have you encountered?

**Comment by sternk** [comment]
Some files still don't go through and the implementation for the symbols
caused additional issues.

**Comment by eugenk** [comment]
How much time do you think you need to finish this? I will need THF support it by Thursday, otherwise there won't be enough time for the evaluation.

**Comment by sternk** [comment]
Should be done this evening. Most files work already.

**Comment by sternk** [comment]
I believe everything except THF-embedded-in-Het/SYN/SYN000-2.p.het works now. This file contains some syntax tests which are not really needed. I'll test all files again (should be done sometime tomorrow) but now all should be good

**Comment by tillmo** [comment]
great!

**Comment by sternk** [comment]
Ok. No new issues turned up. I believe we can close this issue and #1536 now and merge the branch

**Comment by tillmo** [comment]
good, many thanks for jumping in in this situation.

**Comment by eugenk** [comment]
There is still one file that doesn't parse. I don't know if it's a real syntax error or an error in Hets.

It's no big deal for my evaluation, though.

Here's the file: https://gist.github.com/eugenk/e5f9c19df62c1249c8c8/raw/24b1758ff0a67df73ca6621816628fdd3cb12349/SYN000-2.het

**Comment by sternk** [comment]
It's a syntax error due to the parser not supporting the features tested in that file. Currently the effort to extend the parser does not seem to be worth it.

**Comment by cbenzmueller** [comment]
Yes, this file contains some constructs which are not yet relevant. So if
you can parse everything else, then you should be fine.

Best, C.

On Saturday, 5 December 2015, sternk <notifications@github.com> wrote:

> Closed #1544 <https://github.com/spechub/Hets/issues/1544>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spechub/Hets/issues/1544#event-483118510>.
>


-- 
Sent from Gmail Mobile

**Comment by sternk** [comment]
Thanks for the input!
